### DISCUSSION TITLE
Planning a verifier

### DISCUSSION THREAD

**Discussion by hydroper** [airsdk/Adobe-Runtime-Support#2928] [Q&A]
I had a verifier for [a language](https://github.com/violetscript/violetc), inside a .NET environment. I would like to develop one for ActionScript 3 in the Rust language, so that we can get better ASDoc support and other facilities. For now I am not planning the support for Flex (MXML) as I am not familiar with it, but once the goal compiler gets done, maybe...

I am planning to support generics in ActionScript 3, which map type parameters to `*` in the output ABC.

```as3
function f.<T>(cf: (a: Number) => T): void {}
f.<Number> (a => 10);
f          (a => 10); // Not supported on my previous verifier as it omits `.<T>()`
```

Should the third line be supported? Is it important for cases such as `Vector#map`?

I have not yet started the verifier since I will use a little different resolution. And depending on your feedback, the type model might have to be different, not sure...

**Comment by ajwfrost** [comment]
Just wondering whether a better starting point for this would be the ActionScript compiler that's maintained by the Apache Royale project? This is able to convert AS (and MXML) into SWF files as well as JavaScript, and is what we were looking at adopting (or maybe adapting) as the main compiler for AIR SDK...
https://github.com/apache/royale-compiler

**Reply by hydroper** [reply]
Yeah, I wanted to use it. I myself tried to fork and touch a bit on the Royale compiler and no luck as I got JAR errors in the Java source when using VSCode. I myself don't have much understanding of where to go to contribute changes to the compiler. If the actual Maven's Java project were not this barrier, it would be easier :(

**Reply by hydroper** [reply]
Another thing to consider is that destructuring may be difficult to support in Royale since the compiler uses a single binding in the place of destructuring.
